started working on PHP 7 support

tsteur opened the pull request:
refs #8689

This is only partial PHP 7 support. More fixes will be needed.

There is a big blocker re `usort`. It will be hard to make system and UI tests working under PHP5.X and PHP 7.X see https://github.com/piwik/piwik/issues/8689#issuecomment-137424173 . Ideas?

I also created https://github.com/piwik/component-decompress/issues/4

I believe once we fixed the sort problem and the decompress we will have it almost done

tsteur committed:
refs #8689 started working on PHP 7 support

tsteur committed:
make sure to not end in a never ending loop

tsteur commented on .travis.yml:
```diff
@@ -93,8 +103,8 @@ before_install:
 
-  - export GENERATE_TRAVIS_YML_COMMAND="php ./tests/travis/generator/main.php generate:travis-yml --core --verbose"
-  - '[[ "$TRAVIS_JOB_NUMBER" != *.1 || "$TRAVIS_PULL_REQUEST" != "false" ]] || ./tests/travis/autoupdate_travis_yml.sh'
+#  - export GENERATE_TRAVIS_YML_COMMAND="php ./tests/travis/generator/main.php generate:travis-yml --core --verbose"
```
I can enable it again but this means we need to update travis to maybe also run tests for plugins on PHP 7?

diosmosis replied:
You could regenerate the core .travis.yml w/ `--php-versions=7,5.6,5.3.3`

tsteur replied:
We'll need to merge https://github.com/piwik/travis-scripts/pull/9 soon. That should be fine right? Or will it trigger to many auto updates etc? @diosmosis

tsteur commented on core/Db/Adapter.php:
```diff
@@ -39,7 +39,13 @@ public static function factory($adapterName, & $dbInfos, $connect = true)
+        foreach ($dbInfos as $key => $val) {
+            $infos[$key] = $val;
+        }
+
```
each time a new database instance was created we ended up having eg instead of `127.0.0.1` the next time `host=127.0.0.1`, the next time `host=host=127.0.0.1`, ...

tsteur commented on core/Plugin/Dimension/VisitDimension.php:
```diff
@@ -319,20 +319,53 @@ public static function getAllDimensions()
+     * @param VisitDimension[] $dimensions
      */
-    public static function sortByRequiredFields($a, $b)
+    public static function sortDimensions($dimensions)
```
The sort result for PHP7 was different but this way I actually discovered a bug that we did not resolve required fields correctly

sgiehl commented:
I've had a look at the changes... Did you try to (un)install / (de)activate plugins using php7? That doesn't work for me. I tried to figure out why, seems the config is not updated...

tsteur commented:
I haven't tested that yet. Neither the installation (although we have tests for that). It's just a start so far, it will be much more work I think to make it actually compatible and will need much more testing. It's just a start to make the test run and to fix some things that were covered by tests

tsteur commented:
I can confirm activate / deactivate doesn't work. I tried to debug but I couldn't use the interactive debugger so it was hard to debug. I really do not understand the code in `IniFileChain` class even though I tried for quite a while. The problem is `dumpConfig()` returns `null`. I think it's not a problem of that method itself but of some references and because of some changes before. The solution will be probably something like https://github.com/piwik/piwik/pull/8706/files#diff-359a72f49802c7978351b7d4c8c1fe8cR43 for config somewhere. Ideally we would rewrite `Config` and remove all those references etc and clean up that code of `IniFailChain`. I will mention this point in the original issue and we can fix it in a separate pull request. I think we should work on this thing first as maybe some tests fail because of that

tsteur committed:
fix saving config does not work, we need to remove a reference

BTW: The method name Config::forceSave is kinda wrong as it does not
force anything, it saves only if there is a change

tsteur commented:
Activating / deactivating plugin should work now

sgiehl commented:
I can confirm that (de)activating plugins works with the latest changes :+1:

mattab commented:
Targeting 2.15.0 - if we don't merge it for 2.15.0 we should in a separate PR include the  `composer.lock` update

diosmosis commented:
Started looking at this, but won't be able to finish a review for a day or two.

mattab commented:
Fyi: the system test failure should be fixed after a rebase.  LGTM

tsteur committed:
remove php 7 builds as they do not succeed anyway

hpvd commented:
just a question on this:
This Issue is assigned to 2.15 milestone but it has the not in changelog lable.

=> will 2.15 work fine on php 7.0?

If yes, the speed gains should be promoted greatly (if there are any)

On other platforms, php 7.0 is a huge step:
http://www.zend.com/en/resources/php7_infographic

tsteur commented:
It's not fully compatible yet. I presume that's why it is not mentioned there. It might work fun but there could be some problems. I reckon there will be most likely a minor update for PHP 7 once it is released in a few weeks.

hpvd commented:
this issue is marked as done in https://github.com/piwik/piwik/issues/8689
